components/core/src/clp/ffi/ir_stream/Serializer.cpp

```diff
@@ -1,22 +1,221 @@
+ */
+template <typename encoded_variable_t>
+[[nodiscard]] auto serialize_value_array(msgpack::object const& val, vector<int8_t>& buf) -> bool;
+
+auto get_schema_tree_node_type_from_msgpack_val(msgpack::object const& val
+) -> optional<SchemaTreeNode::Type> {
+    optional<SchemaTreeNode::Type> ret_val;
+    switch (val.type) {
+        case msgpack::type::POSITIVE_INTEGER:
+        case msgpack::type::NEGATIVE_INTEGER:
+            ret_val.emplace(SchemaTreeNode::Type::Int);
+            break;
+        case msgpack::type::FLOAT32:
+        case msgpack::type::FLOAT:
+            ret_val.emplace(SchemaTreeNode::Type::Float);
+            break;
+        case msgpack::type::STR:
+            ret_val.emplace(SchemaTreeNode::Type::Str);
+            break;
+        case msgpack::type::BOOLEAN:
+            ret_val.emplace(SchemaTreeNode::Type::Bool);
+            break;
+        case msgpack::type::NIL:
+        case msgpack::type::MAP:
+            ret_val.emplace(SchemaTreeNode::Type::Obj);
+            break;
+        case msgpack::type::ARRAY:
+            ret_val.emplace(SchemaTreeNode::Type::UnstructuredArray);
+            break;
+        default:
+            return std::nullopt;
+    }
+    return ret_val;
+}
+
+auto serialize_empty_object(vector<int8_t>& buf) -> void {
+    buf.push_back(cProtocol::Payload::ValueEmpty);
+}
+
+auto serialize_value_int(int64_t val, vector<int8_t>& buf) -> void {
+    if (INT8_MIN <= val && val <= INT8_MAX) {
+        buf.push_back(cProtocol::Payload::ValueInt8);
+        buf.push_back(static_cast<int8_t>(val));
+    } else if (INT16_MIN <= val && val <= INT16_MAX) {
+        buf.push_back(cProtocol::Payload::ValueInt16);
+        serialize_int(static_cast<int16_t>(val), buf);
+    } else if (INT32_MIN <= val && val <= INT32_MAX) {
+        buf.push_back(cProtocol::Payload::ValueInt32);
+        serialize_int(static_cast<int32_t>(val), buf);
+    } else {  // (INT64_MIN <= val && val <= INT64_MAX)
+        buf.push_back(cProtocol::Payload::ValueInt64);
+        serialize_int(val, buf);
+    }
+}
+
+auto serialize_value_float(double val, vector<int8_t>& buf) -> void {
+    buf.push_back(cProtocol::Payload::ValueFloat);
+    serialize_int(bit_cast<uint64_t>(val), buf);
+}
+
+auto serialize_value_bool(bool val, vector<int8_t>& buf) -> void {
+    buf.push_back(val ? cProtocol::Payload::ValueTrue : cProtocol::Payload::ValueFalse);
+}
+
+auto serialize_value_null(vector<int8_t>& buf) -> void {
+    buf.push_back(cProtocol::Payload::ValueNull);
+}
+
+template <typename encoded_variable_t>
+auto serialize_value_string(string_view val, vector<int8_t>& buf) -> bool {
+    if (string_view::npos == val.find(' ')) {
+        return serialize_string_packet(val, buf);
+    }
+    return serialize_clp_string<encoded_variable_t>(val, buf);
+}
+
+template <typename encoded_variable_t>
+auto serialize_value_array(msgpack::object const& val, vector<int8_t>& buf) -> bool {
+    std::ostringstream oss;
```

Comment:
nitpick: any possibility we could re-use the same ostringstream object instead of recreating it every time this function is called?

Reply:
We will deprecate the current implementation whenever this PR gets merged: #465